extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -14,43 +14,407 @@
         `Experiment ${EXPERIMENT} is not turned on.`);
-    this.container_ = this.element.ownerDocument.createElement('div');
-    this.container_.textContent = this.myText_;
-    this.element.appendChild(this.container_);
-    this.applyFillContent(this.container_, /* replacedContent */ true);
+
+    this.inlineData_ = this.getInlineData_();
+
+    const inputElements = childElementsByTag(this.element, 'INPUT');
+    userAssert(inputElements.length === 1,
+        `${TAG} should contain exactly one <input> child`);
+    this.inputElement_ = inputElements[0];
+
+    this.filter_ = userAssert(this.element.getAttribute('filter'),
+        `${TAG} requires "filter" attribute.`);
+    userAssert(isEnumValue(FilterType, this.filter_),
+        `Unexpected filter: ${this.filter_}`);
+
+    this.minChars_ = this.element.hasAttribute('min-characters') ?
+      parseInt(this.element.getAttribute('min-characters'), 10) : 1;
+    this.maxEntries_ = parseInt(this.element.getAttribute('max-entries'), 10);
+
+    return this.mutateElement(() => {
+      this.container_ = this.createContainer_();
+      this.element.appendChild(this.container_);
+    });
+  }
+
+  /**
+   * Reads the 'items' data from the child <script> element.
+   * For use with static local data.
+   * @return {?Array}
+   * @private
+   */
+  getInlineData_() {
+    const scriptElements = childElementsByTag(this.element, 'SCRIPT');
+    if (!scriptElements.length) {
+      return null;
+    }
+    userAssert(scriptElements.length === 1,
+        `${TAG} should contain at most one <script> child`);
+    const scriptElement = scriptElements[0];
+    userAssert(isJsonScriptTag(scriptElement),
+        `${TAG} should be inside a <script> tag with type="application/json"`);
+    const json = tryParseJson(scriptElement.textContent,
+        error => {
+          throw error;
+        });
+    return json['items'] ? json['items'] : [];
+  }
+
+  /**
+   * Creates and returns <div> that contains the template-rendered children.
+   * @return {Element}
+   * @private
+   */
+  createContainer_() {
+    const container = this.element.ownerDocument.createElement('div');
+    container.classList.add('i-amphtml-autocomplete-results');
+    container.setAttribute('role', 'list');
+    toggle(container, false);
+    this.applyFillContent(container, /* replacedContent */ true);
+    return container;
   }
 
   /** @override */
   layoutCallback() {
-    // Actually load your resource or render more expensive resources.
-    return Promise.resolve();
+    // Disable autofill in browsers.
+    this.inputElement_.setAttribute('autocomplete', 'off');
+
+    // No static local data to filter against.
+    if (!this.inlineData_) {
+      return Promise.resolve();
+    }
+
+    // Register event handlers.
+    this.inputElement_.addEventListener('input',
+        this.inputHandler_.bind(this));
+    this.inputElement_.addEventListener('keydown',
+        this.keyDownHandler_.bind(this));
```

Comment:
You can use a fat-arrow function to fix the type check error. It looks like it want the function passed to addEventListener to have particular return values.

I would say keep the Promise return values in these, since they make it easier to test the code, and use a fat-arrow:
```js
this.inputElement_.addEventListener('keydown', e => this. keyDownHandler_(e));
```

Reply:
Thanks for this advice!!